rts/motoko-rts/src/custom.rs

```diff
@@ -0,0 +1,326 @@
+    unsafe fn into_value(self, mem: &mut impl Memory) -> MotokoResult<Value> {
+        let vec = self.0;
+        let value = alloc_blob(mem, Bytes(vec.len() as u32));
+        let blob = value.as_blob_mut();
+        let mut dest = blob.payload_addr();
+        for item in vec.into_iter() {
+            *dest = item;
+            dest = dest.add(1);
+        }
+        Ok(allocation_barrier(value))
+        // ^^ Is this allocation barrier necessary?
+    }
+}
+
+impl<T: FromValue> FromValue for Vec<T> {
+    unsafe fn from_value(value: Value, mem: &mut impl Memory) -> MotokoResult<Self> {
+        match value.tag() {
+            TAG_ARRAY => {
+                let array = value.as_array();
+                let len = array.len();
+                (0..len)
+                    .into_iter()
+                    .map(|i| T::from_value(array.get(i), mem))
+                    .collect()
+            }
+            tag => Err(MotokoError::UnexpectedTag(tag)),
+        }
+    }
+}
+impl<T: IntoValue> IntoValue for Vec<T> {
+    unsafe fn into_value(self, mem: &mut impl Memory) -> MotokoResult<Value> {
+        let value = alloc_array(mem, self.len() as u32);
+        let array = value.as_array();
+        let mut dest = array.payload_addr();
+        for item in self.into_iter() {
+            *dest = item.into_value(mem)?;
+            dest = dest.add(1);
+        }
+        Ok(allocation_barrier(value))
+    }
+}
+
+impl FromValue for () {
+    unsafe fn from_value(_value: Value, _mem: &mut impl Memory) -> MotokoResult<Self> {
+        Ok(())
+    }
+}
+impl IntoValue for () {
+    unsafe fn into_value(self, _mem: &mut impl Memory) -> MotokoResult<Value> {
+        Ok(unit())
+    }
+}
+
+// Implement `FromValue` and `IntoValue` for tuples
+#[tuple_macro]
+macro_rules! tuple_impl {
+    ($len:expr; $($name:ident = $index:tt),+) => {
+        impl<$($name: FromValue),+> FromValue for ($($name),+) {
+            unsafe fn from_value(value: Value, mem: &mut impl Memory) -> MotokoResult<Self> {
+                const LENGTH: u32 = $len;
+                match value.tag() {
+                    TAG_ARRAY => {
+                        let array = value.as_array();
+                        assert_eq!(array.len(), LENGTH, "Unexpected tuple length");
+                        Ok((
+                            $($name::from_value(array.get($index), mem)?),+
+                        ))
+                    }
+                    tag => Err(MotokoError::UnexpectedTag(tag)),
+                }
+            }
+        }
+        impl<$($name: IntoValue),+> IntoValue for ($($name),+) {
+            unsafe fn into_value(self, mem: &mut impl Memory) -> MotokoResult<Value> {
+                const LENGTH: u32 = $len;
+                let value = alloc_array(mem, LENGTH);
+                let array = value.as_array();
+                let dest = array.payload_addr();
+                $(*(dest.add($index)) = $name::into_value(self.$index, mem)?;)+
```

Comment:
Here, we would have to use `Array.initialize()` as otherwise a initialization barrier may be missing (resolving forwarding pointers during incremental update phase).